### PR TITLE
PgLwLock implementation

### PR DESCRIPTION
**Pull request by jamessewell**
This is the start of the shmem story

### CODE REVIEW

File: pgx/src/shmem.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        unsafe { (*self.inner.get()).as_ref().unwrap().share() }
+    }
+
+    pub fn exclusive(&self) -> PgLwLockExclusiveGuard<T> {
+        unsafe { (*self.inner.get()).as_mut().unwrap().exclusive() }
+    }
+
+    // To use a lock it has to be attached to an allocated Postgres LWLock, which
+    // we look up by name
+    pub fn attach(&self, lock: String, value: T) -> Result<(), T> {
```
Comment (eeeebbbbrrrr) [inline pgx/src/shmem.rs:26]: why not just make this the `new` constructor instead?

**Comment by eeeebbbbrrrr** [comment]
I'd like to see doc comments for all the `pub` stuff here.  Even if just copy-pasta from the Postgres sources

**Comment by eeeebbbbrrrr** [comment]
also, a few `#[pg_test]` functions in the `./pgx/pgx-tests/src/tests/` directory would be nice too.  Even if they're just ensuring the various `PgLwLock` methods don't segfault, that'd be better than nothing.  I don't think the pg_test framework is capable enough to do cross-process testing, but maybe you can use the `pg_locks` view to see if a LWLock has been acquired (do they show up there?)?

File: pgx/src/shmem.rs (hunk 1 above)
Comment (jamessewell) [reply pgx/src/shmem.rs:44]: Generally (at least my usage so far) the lock gets created (but not attached) as a global, then attached later on.

let me think about this a bit more

File: pgx/src/shmem.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+/// This lock can not be poisoned from Rust. Panic and Abort are handled by
+/// PostgreSQL cleanly.
+
+pub struct PgLwLock<T> {
+    inner: UnsafeCell<Option<PgLwLockInner<T>>>,
+}
+
+impl<T> PgLwLock<T> {
+    /// Create a new lock for T by attaching a LWLock, which is looked up by name
+    pub fn new(lock: String, value: T) -> Self {
```
Comment (zombodb) [inline pgx/src/shmem.rs:29]: Any particular reason why the lock name throughout is a `String` and not a `&str`?  

Maybe it should even be `&'static str`?

**Comment by jamessewell** [comment]
Look it should probably be a cstr - then that removes an Err case

**Comment by eeeebbbbrrrr** [comment]
That make sense if we want to expose CStr to users?  I guess I lean towards &’static str tho. 

Is it expected (or desired) for lock names to be dynamically generated or are they typically just static values anyways?

**Comment by jamessewell** [comment]
Typically static

**Comment by zombodb** [comment]
This is really great. Awesome work. I’ll get it merged soon. Thinking about what some docs should say

**Comment by eeeebbbbrrrr** [comment]
wow, what just happened?  I merged this.  That's what!

**Pull request merged**